magento/magento2#24186: CLI: cron:install. Break tasks to default and non-optional.

atwixfirster opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

`bin/magento cron:install` command adds the extra cron job - `update/cron.php`:

```
#~ MAGENTO START c5f9e5ed71cceaabc4d4fd9b3e827a2b
#* * * * * /usr/bin/php7.1 /home/dev/sites/pgs/bin/magento cron:run 2>&1 | grep -v "Ran jobs by schedule" >> /home/dev/sites/pgs/var/log/magento.cron.log
#* * * * * /usr/bin/php7.1 /home/dev/sites/pgs/update/cron.php >> /home/dev/sites/pgs/var/log/update.cron.log
#* * * * * /usr/bin/php7.1 /home/dev/sites/pgs/bin/magento setup:cron:run >> /home/dev/sites/pgs/var/log/setup.cron.log
#~ MAGENTO END c5f9e5ed71cceaabc4d4fd9b3e827a2b
```


### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento2#24186: bin/magento cron:install adds extra cron job
2. https://github.com/magento/magento2/issues/10040
3. https://github.com/magento/devdocs/issues/1740
### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
**Example #1**
1. Install Magento 2.3 via `bin/magento setup:install` command
2. Run `bin/magento cron:install`
3. Run `crontab -l`
4. Check `<magento root folder>/update` folder. 

**Actual result:**
1. `<magento root folder>/update` folder does not exist.
2. `<magento root folder>/update/cron.php` does not exist.

**Example #2**
1. Install Magento 2.3 via `Web Setup Wizard`
2. Run `bin/magento cron:install`
3. Run `crontab -l`
4. Check `<magento root folder>/update` folder. 

**Actual result:**
1. `<magento root folder>/update` folder does not exist.
2. `<magento root folder>/update/cron.php` does not exist.

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

CC: @hostep

hostep commented:
Hi @atwixfirster, thanks for taking this on!

First things first, the file `update/cron.php` actually does exists. If you install Magento using composer, or [download the archive](https://magento.com/tech-resources/download), in both cases the file exists.
In the case of composer, the directory `update` and all the files in there are only created when you run `composer create-project`, however if you only run `composer install` you'll not see the `update` directory getting created (if it didn't exists before).

The reason why you perhaps do not have it in your case, is that you might have ignored the `update` directory in git, because you probably don't need it (that's also what we are doing, and why I've opened https://github.com/magento/magento2/issues/10040 in the past).
Another case where the `update` directory doesn't exists, is if you simply clone the magento/magento2 repository.

So we need to fix this in some other way, not by simply removing one of those non-essential cron commands (`setup:cron:run` is also a non-essential one in my opinion).

I see a possible solution where we can add an extra attribute in the xml to mark a certain cronjob as "essential" (or the other way around: "optional").
Then, we can add a new flag to the `cron:install` command, where we can say we only want the "essential" tasks to get installed (or the other way around: where we can skip the "optional" tasks).

Does this make sense?
I'm not very good in naming things, so maybe these "essential" and "optional" words I'm using can have a better name...

atwixfirster commented:
Thank you for the detailed explanations, @hostep 👍 

I've added the next cron groups:
- `default` (when customer runs `cron:install` command then all crons from this group will be installed)
- `optional` (crons from this group will be installed when customer runs `cron:install` command with `-o` or `--optional` flag)

So, I've moved `cronMagento` to `default` and `cronUpdate`, `cronSetup` to `optional` groups.

**Test case #1**

```
bin/magento cron:install -o
```

**Expected result**

```
#~ MAGENTO START 4b4b4d53b51a7e3ff2fc59a11338c37ff8cc917d0b5701344c7bd1ca530f2b0c
* * * * * /usr/bin/php7.1 /home/dev/sites/2.3.1/bin/magento cron:run | grep -v "Ran jobs by schedule" >> /home/dev/sites/2.3.1/var/log/magento.cron.log
* * * * * /usr/bin/php7.1 /home/dev/sites/2.3.1/update/cron.php >> /home/dev/sites/2.3.1/var/log/update.cron.log
* * * * * /usr/bin/php7.1 /home/dev/sites/2.3.1/bin/magento setup:cron:run >> /home/dev/sites/2.3.1/var/log/setup.cron.log
#~ MAGENTO END 4b4b4d53b51a7e3ff2fc59a11338c37ff8cc917d0b5701344c7bd1ca530f2b0c
```

**Test case #2**

```
bin/magento cron:install
```

**Expected result**

```
#~ MAGENTO START 4b4b4d53b51a7e3ff2fc59a11338c37ff8cc917d0b5701344c7bd1ca530f2b0c
* * * * * /usr/bin/php7.1 /home/dev/sites/bravo/2.3.1/bin/magento cron:run | grep -v "Ran jobs by schedule" >> /home/dev/sites/bravo/2.3.1/var/log/magento.cron.log
#~ MAGENTO END 4b4b4d53b51a7e3ff2fc59a11338c37ff8cc917d0b5701344c7bd1ca530f2b0c
```

If this solution is ok for you?

Thank you!

hostep commented:
Thanks for the update @atwixfirster, looks great at first sight!

But I would reverse the meaning of the extra flag, right now it changes the behavior of the command when running it without extra flags, which is backwards incompatible in my eyes.

If the flag would mean 'Exclude optional crons', then we have backwards compatibility again I think.

hostep commented on app/code/Magento/Cron/Console/Command/CronInstallCommand.php:
```diff
@@ -76,4 +87,25 @@ protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $defaultCrons = [];
+
+        if (count($crons)) {
```
This `count` check is not necessary. We can be sure the variable is an array as its enforced by the typehint in the argument. And if the array is empty, the foreach will not do anything. So you can drop the `count` check here.

atwixfirster replied:
**done**

Thanks

atwixfirster commented:
> Thanks for the update @atwixfirster, looks great at first sight!
> 
> But I would reverse the meaning of the extra flag, right now it changes the behavior of the command when running it without extra flags, which is backwards incompatible in my eyes.
> 
> If the flag would mean 'Exclude optional crons', then we have backwards compatibility again I think.

So, your final suggestion:
- install all crons per `bin/magento cron:install`
- add a flag which allow to install `cronMagento` only?

Or leave a current solution: when we install `cronUpdate`, `cronSetup` per `bin/magento cron:install -o` command?

Thank you! :)

hostep commented:
Yes correct:
- install all crons when not providing the new flag.
- and install only the non-optional one when providing the new flag.

Then we don't have a change in behavior compared to how it was before.

atwixfirster commented:
> Yes correct:
> 
>     * install all crons when not providing the new flag.
> 
>     * and install only the non-optional one when providing the new flag.
> 
> 
> Then we don't have a change in behavior compared to how it was before.

**implemented**

So, final scenarios:

**Test case 1**

```
bin/magento cron:install
```

**Expected result**

```
#~ MAGENTO START 4b4b4d53b51a7e3ff2fc59a11338c37ff8cc917d0b5701344c7bd1ca530f2b0c
* * * * * /usr/bin/php7.1 /home/dev/sites/2.3.1/bin/magento cron:run | grep -v "Ran jobs by schedule" >> /home/dev/sites/2.3.1/var/log/magento.cron.log
* * * * * /usr/bin/php7.1 /home/dev/sites/2.3.1/update/cron.php >> /home/dev/sites/2.3.1/var/log/update.cron.log
* * * * * /usr/bin/php7.1 /home/dev/sites/2.3.1/bin/magento setup:cron:run >> /home/dev/sites/2.3.1/var/log/setup.cron.log
#~ MAGENTO END 4b4b4d53b51a7e3ff2fc59a11338c37ff8cc917d0b5701344c7bd1ca530f2b0c
```

**Test case 2**

```
bin/magento cron:install -d
```

**Expected result**

```
#~ MAGENTO START 4b4b4d53b51a7e3ff2fc59a11338c37ff8cc917d0b5701344c7bd1ca530f2b0c
* * * * * /usr/bin/php7.1 /home/dev/sites/bravo/2.3.1/bin/magento cron:run | grep -v "Ran jobs by schedule" >> /home/dev/sites/bravo/2.3.1/var/log/magento.cron.log
#~ MAGENTO END 4b4b4d53b51a7e3ff2fc59a11338c37ff8cc917d0b5701344c7bd1ca530f2b0c
```

**Test case 3**

```
bin/magento cron:install --non-optional
```

**Expected result**

```
#~ MAGENTO START 4b4b4d53b51a7e3ff2fc59a11338c37ff8cc917d0b5701344c7bd1ca530f2b0c
* * * * * /usr/bin/php7.1 /home/dev/sites/bravo/2.3.1/bin/magento cron:run | grep -v "Ran jobs by schedule" >> /home/dev/sites/bravo/2.3.1/var/log/magento.cron.log
#~ MAGENTO END 4b4b4d53b51a7e3ff2fc59a11338c37ff8cc917d0b5701344c7bd1ca530f2b0c
```

Thank you, @hostep !

hostep commented:
Nice, good job!

The last little comment I have is your commit message, that one no longer represent what this PR is achieving, can you amend your commit message and force push again?

I'll give back control to @dmytro-ch now, from my point of view this looks good!

atwixfirster commented:
> The last little comment I have is your commit message, that one no longer represent what this PR is achieving, can you amend your commit message and force push again?

@hostep , I squashed all commits and looks like I can not revert back all commits.

😞

hostep commented:
No I mean you can just update the commit message, something like this:

`git commit --amend`

This will open an editor to change your commit message.

After that you can force push your branch it again.

atwixfirster committed:
magento/magento2#24186: CLI: cron:install. Break tasks to default and non-optional.

atwixfirster commented:
> No I mean you can just update the commit message, something like this:
> 
> `git commit --amend`
> 
> This will open an editor to change your commit message.
> 
> After that you can force push your branch it again.

**done**

Thank you, @hostep

engcom-Delta commented:
:heavy_check_mark: QA passed

sidolov committed:
Make constants private

lbajsarowicz commented:
OK. I was missing the information on devdocs.

hostep commented:
@lbajsarowicz: that's a good point!

@atwixfirster: I haven't been following devdocs for the last couple of months, is there already documentation added for this new flag? (it might be on the 2.3.4-develop branch of devdocs which hasn't been published yet, but I haven't checked, so I'm not sure)

atwixfirster commented:
@lbajsarowicz

> OK. I was missing the information on devdocs.

PR was created: https://github.com/magento/devdocs/pull/5298 and @hostep has approved it :)

I did all from my side guys :)

lbajsarowicz commented:
Awesome job @atwixfirster :+1:

atwixfirster commented:
> Awesome job @atwixfirster 👍

Thanks, @lbajsarowicz 

But I did not find my PR in the develop branches. So, I've created a new one - https://github.com/magento/devdocs/pull/6428 and mentioned you guys.

Thanks!
